### PR TITLE
feat: Change icon used in Collapsible widget

### PR DESCRIPTION
**Pull request by ppwadhwa**
This PR adds the ability to change the icon on the toggle QPushButton.  Currently, the icon is a part of the text string.  This PR will change it so that an icon is set on the button.  The original icon is still used if there isn't one set, however, the default string is used to create an icon.  

You can now instantiate a collapsible widget while also declaring an `expandedIcon` and/or `collapsedIcon`.  These icons can be a QIcon or a text string.

```
icon1 = icon(FA5S.smile, color='white')
icon2 = icon(FA5S.frown, color='white')
collapsible = QCollapsible("Advanced analysis", expandedIcon=icon1, collapsedIcon=icon2)
```
Screen Shot 2022-11-21 at 4 14 03 PM
Screen Shot 2022-11-21 at 4 14 11 PM

```
collapsible = QCollapsible("Advanced analysis", expandedIcon='-', collapsedIcon="+")
```
Screen Shot 2022-11-21 at 4 16 53 PM
Screen Shot 2022-11-21 at 4 16 46 PM

### CODE REVIEW

File: src/superqt/collapsible/_collapsible.py [hunk 1]
```diff
@@ -63,6 +72,48 @@ def content(self) -> QWidget:
         """Return the current content widget."""
         return self._content
 
+    def set_expanded_icon(self, icon=None):
```
Comment (goanpeca) [inline src/superqt/collapsible/_collapsible.py:75]: This should follow the Qt naming convention so, `setExpandedIcon`

File: src/superqt/collapsible/_collapsible.py [hunk 2]
```diff
@@ -11,16 +20,16 @@ class QCollapsible(QFrame):
     Based on https://stackoverflow.com/a/68141638
     """
 
-    _EXPANDED = "▼  "
-    _COLLAPSED = "▲  "
-
     def __init__(self, title: str = "", parent: Optional[QWidget] = None):
```
Comment (goanpeca) [inline src/superqt/collapsible/_collapsible.py:23]: This should maybe also expose the `collapsedIcon` and the `expandedIcon` as Optional arguments (as some buttons allow in Qt)?

What do you think @tlambert03 ?

File: src/superqt/collapsible/_collapsible.py [hunk 3]
```diff
@@ -63,6 +72,48 @@ def content(self) -> QWidget:
...
+            self._EXPANDED = icon
+        else:
+            pixmap = QPixmap(16, 16)
+            pixmap.fill(Qt.transparent)
+            painter = QPainter(pixmap)
+            color = self._toggle_btn.palette().color(QPalette.WindowText)
+            painter.setPen(color)
+            symbol = "▼"
+            painter.drawText(QRect(0, 0, 16, 16), Qt.AlignCenter, symbol)
+            painter.end()
```
Comment (goanpeca) [inline src/superqt/collapsible/_collapsible.py:88]: All this code is duplicated so we could use a helper private function which just receives the symbol. 

```python
def _convert_symbol_to_icon(self, symbol: str) -> QIcon:
    pixmap = QPixmap(16, 16)
    pixmap.fill(Qt.transparent)
    painter = QPainter(pixmap)
    color = self._toggle_btn.palette().color(QPalette.WindowText)
    painter.setPen(color)
    painter.drawText(QRect(0, 0, 16, 16), Qt.AlignCenter, symbol)
    painter.end()
    return QIcon(pixmap)
```

---

@tlambert03 do we want to support symbols as text in general or limiting the approach to icons only is ok?

**Review by goanpeca — CHANGES_REQUESTED** [review]
Hi @ppwadhwa thanks for working on this! Great work 🚀 

Added some comments to clean up the code a bit :)

File: src/superqt/collapsible/_collapsible.py (hunk 3 above)
Comment (goanpeca) [reply src/superqt/collapsible/_collapsible.py:88]: Also not sure how the icon size, will play out with custom QSS 🤔

Comment (ppwadhwa) [reply src/superqt/collapsible/_collapsible.py:88]: I just know that this doesn't seem to work unless I have a size in there.  16 seemed to look the best.  How would you suggest I handle this?

Comment (goanpeca) [reply src/superqt/collapsible/_collapsible.py:88]: Maybe get the size of the QPushButton or the default font size and use that

File: src/superqt/collapsible/_collapsible.py (hunk 2 above)
Comment (ppwadhwa) [reply src/superqt/collapsible/_collapsible.py]: I added this.

File: src/superqt/collapsible/_collapsible.py (hunk 3 above)
Comment (ppwadhwa) [reply src/superqt/collapsible/_collapsible.py]: ok. will look at doing that.

File: tests/test_collapsible.py [hunk 4]
```diff
@@ -84,4 +84,4 @@ def test_changing_text(qtbot):
     wdg = QCollapsible()
     wdg.setText("Hi new text")
     assert wdg.text() == "Hi new text"
-    assert wdg._toggle_btn.text() == QCollapsible._COLLAPSED + "Hi new text"
+    assert wdg._toggle_btn.text() == "Hi new text"
```
[the remark this answers is not in the document]
Comment (ppwadhwa) [reply tests/test_collapsible.py:90]: done!

File: tests/test_fonticon/test_collapsible.py [hunk 5]
```diff
@@ -1,9 +1,32 @@
...
 
 from superqt import QCollapsible
+from superqt.fonticon import icon
+from superqt.fonticon._qfont_icon import QFontIconStore
+
+TEST_PREFIX = "ico"
+TEST_CHARNAME = "smiley"
+TEST_CHAR = "\ue900"
+TEST_GLYPHKEY = f"{TEST_PREFIX}.{TEST_CHARNAME}"
+FONT_FILE = Path(__file__).parent / "icontest.ttf"
```
Comment (tlambert03) [inline tests/test_fonticon/test_collapsible.py:16]: please don't duplicate all this code from `test_fonticon.py`.  If you need these fixtures here as well, place them somewhere that they can be used in both files

File: src/superqt/collapsible/_collapsible.py [hunk 6]
```diff
@@ -63,6 +78,41 @@ def content(self) -> QWidget:
...
+            self._EXPANDED = icon
+        elif icon and isinstance(icon, str):
+            self._EXPANDED = self._convert_symbol_to_icon(icon)
+        else:
+            symbol = "▼"
+            self._EXPANDED = self._convert_symbol_to_icon(symbol)
+
+    def setCollapsedIcon(self, icon: Optional[Union[QIcon, str]] = None):
+        """Set the icon on the toggle button when the widget is collapsed."""
+
```
Comment (goanpeca) [inline src/superqt/collapsible/_collapsible.py:105]: Lets remove this empty line :)

File: src/superqt/collapsible/_collapsible.py [hunk 7]
```diff
@@ -63,6 +78,41 @@ def content(self) -> QWidget:
...
+        painter = QPainter(pixmap)
+        color = self._toggle_btn.palette().color(QPalette.WindowText)
+        painter.setPen(color)
+        painter.drawText(QRect(0, 0, size, size), Qt.AlignCenter, symbol)
+        painter.end()
+        return QIcon(pixmap)
+
+    def setExpandedIcon(self, icon: Optional[Union[QIcon, str]] = None):
+        """Set the icon on the toggle button when the widget is expanded."""
+
```
Comment (goanpeca) [inline src/superqt/collapsible/_collapsible.py:94]: Lets remove this empty line :)

**Review by goanpeca — CHANGES_REQUESTED** [review]

File: src/superqt/collapsible/_collapsible.py [hunk 8]
```diff
@@ -21,18 +23,27 @@ class QCollapsible(QFrame):
     Based on https://stackoverflow.com/a/68141638
     """
 
-    _EXPANDED = "▼  "
-    _COLLAPSED = "▲  "
+    _expanded_icon = "▼  "
+    _collapsed_icon = "▲  "
```
Comment (goanpeca) [inline src/superqt/collapsible/_collapsible.py:27]: Why do we need to keep them if they are hardcoded in the methods?

**Review by goanpeca — APPROVED** [review]
Thanks for working on this @ppwadhwa, works great 🚀

**Comment by tlambert03** [comment]
good stuff.  thanks both!

**Review by tlambert03 — APPROVED** [review]

**Comment by goanpeca** [comment]
Thanks for the quick review 🤭 

@tlambert03 do you think you could make a new release with the latest changes sometime soon :) ?

Thanks for the awesome project!

**Comment by tlambert03** [comment]
Yep 👍 will do tomorrow

**Comment by ppwadhwa** [comment]
thanks @tlambert03 !!!

**Comment by tlambert03** [comment]
v0.4.1 building now... should be up soon

**Pull request merged**